source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc

```diff
@@ -20,29 +23,72 @@ Extensions::Common::Aws::Signer& FilterConfigImpl::signer() { return *signer_; }
 FilterStats& FilterConfigImpl::stats() { return stats_; }
 
 const std::string& FilterConfigImpl::hostRewrite() const { return host_rewrite_; }
+bool FilterConfigImpl::useUnsignedPayload() const { return use_unsigned_payload_; }
 
 FilterStats Filter::generateStats(const std::string& prefix, Stats::Scope& scope) {
   const std::string final_prefix = prefix + "aws_request_signing.";
   return {ALL_AWS_REQUEST_SIGNING_FILTER_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
 }
 
-Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
   const auto& host_rewrite = config_->hostRewrite();
+  const bool use_unsigned_payload = config_->useUnsignedPayload();
+
   if (!host_rewrite.empty()) {
     headers.setHost(host_rewrite);
   }
 
+  if (!use_unsigned_payload && !end_stream) {
+    request_headers_ = &headers;
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
   try {
-    config_->signer().sign(headers);
+    ENVOY_LOG(debug, "aws request signing from decodeHeaders use_unsigned_payload: {}",
+              use_unsigned_payload);
+    config_->signer().sign(headers, use_unsigned_payload);
     config_->stats().signing_added_.inc();
   } catch (const EnvoyException& e) {
+    // TODO: sign should not throw to avoid exceptions in the request path
     ENVOY_LOG(debug, "signing failed: {}", e.what());
     config_->stats().signing_failed_.inc();
   }
 
   return Http::FilterHeadersStatus::Continue;
 }
 
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  if (config_->useUnsignedPayload()) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (!end_stream) {
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  decoder_callbacks_->addDecodedData(data, false);
```

Comment:
Hmm , I think this should happen before returning StopIterationAndBuffer.

Reply:
Based on the docstrings (item 2, reference below) and implementation in other filters, I _think_ I've used this correctly (this line is only reachable when `end_stream` is `true`.

https://github.com/envoyproxy/envoy/blob/ad4919dd70276e4d89b011a03376fe6d20475547/include/envoy/http/filter.h#L335-L364

Reply:
Yeah you are right, I got confused with the case where the data will be changed/replaced (e.g.: compression).